cypress-io/cypress
# fix(webpack-dev-server): touch component-index during onSpecsChange to avoid writing to app file

- Closes #24398
- Closes #24876

### Additional details


In our Webpack plugin, we have a hook that runs when specs change (files are added or removed). When this happens, we use `uTimesSync` to update the metadata for a file in `@cypress/webpack-dev-server`, `browser.js`. This "tricks" webpack into re-compiling the bundle since a dependency has technically changed.

Something changed with MacOS Ventura where we are no longer able to write to files inside of the Cypress application directory without specific application permissions set up in system preferences (unclear what these actually are - worked for some and not for others). This is despite the user having permission to write to the file, according to the file permissions.

One way to avoid this issue is to, rather than modifying `browser.js` to trigger a recompilation of specs, modify the component-index HTML file that exists in the project. The component index is a good candidate for this because:

1. We can count on this file to exist because it needs to in order for CT to work.
2. It should always be a dependency of the bundle (so it will trigger a re-compile)
3. We should always be able to write to it (if we don't have permission to write to the project directory I think we have bigger issues).

### Steps to test


1. Open a project with CT
2. Run a spec
4. Create a new spec (with create from component flow or something)
5. Try to run that spec
6. Verify that the spec runs correctly (this means that webpack re-compiled)

The bug is a little bit harder to reproduce since you need to be running Ventura. I've been able to repro it by running CT with [Real World App](https://github.com/cypress-io/cypress-realworld-app). Any time I edit and save a CT spec a couple of times while in open mode, the error will throw for me. I tested the binary on my machine running Ventura with RWA and it worked 👍🏻 

### How has the user experience changed?


### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
